### ISSUE TITLE
weird behavior when run from a symlinked maildir 

### ISSUE THREAD

**Issue by teto** [elizagamedev/mujmap#11]
I am trying to setup mujmap on my own before checkjing out the homemanager module.
I have my maildir folder as a symlink towards  /mnt/ext/maildir and I think this confuses mujmap
```
✖2 ❯ mujmap -C ~/maildir sync                       
[2022-05-16T22:22:47Z WARN  mujmap::sync] Could not read mujmore state file `/home/teto/maildir/mujmap.state.json': No such file or directory (os error 2)
error: Could not sync mail: Could not open remote session: Could not open session at https://jmap.fastmail.com:443/.well-known/jmap: https://jmap.fastmail.com/session: status code 401 (redirected from https://jmap.fastmail.com/.well-known/jmap)
~/.config took 4s 
✖1 ❯ mujmap -C ~/maildir sync
~/.config 
✖130 ❯ cd ~/maildir
~/maildir 
➜ cp mujmap.toml .notmuch/
~/maildir 
➜ mujmap -C .notmuch sync 
[2022-05-16T22:25:25Z WARN  mujmap::sync] Could not read mujmore state file `.notmuch/mujmap.state.json': No such file or directory (os error 2)
error: Could not sync mail: Could not open local database: Given maildir path `/mnt/ext/maildir/.notmuch' is not a subdirectory of the notmuch root `/home/teto/maildir'
➜ mujmap -C . sync       
[2022-05-16T22:25:57Z WARN  mujmap::sync] Could not read mujmore state file `./mujmap.state.json': No such file or directory (os error 2)
error: Could not sync mail: Could not open local database: Given maildir path `/mnt/ext/maildir' is not a subdirectory of the notmuch root `/home/teto/maildir'
~/maildir 
➜ ls -l maildir 
lrwxrwxrwx 1 teto users 16 2021-10-02  maildir -> /mnt/ext/maildir
```
as seen previously, if I run it from another directory with `-C ~/maildir` it seems to work.

```
mujmap -C maildir sync -v
error: Could not sync mail: Could not open remote session: Could not open session at https://jmap.fastmail.com:443/.well-known/jmap: https://jmap.fastmail.com/session: status code 401 (redirected from https://jmap.fastmail.com/.well-known/jmap)
```

My credentials look ok (btw I can access my mail over jmap with meli https://meli.delivery/#) so I wondered if the newline in my password could be an issue (I run `password_command = "pass-show perso/fastmail_mc"` which returns my password plus a newline apparently). Is there a way to just dump my password in the config file to make sure that's not the cause for the 401.

**Comment by elizagamedev** [comment]
For the symlink portion of the issue, yeah, it seems like mujmap needs more robust handling of symlinks. I'll investigate a solution for that.

As for your credentials issue, there was a bug in 0.1.0 that caused password commands which failed with non-zero return codes to try and authenticate with the (likely blank) stdout anyway. I recommend installing from the latest revision and checking if the error message becomes more useful. If you are already running from the latest revision already, I'm not quite sure what's going on yet.

**Comment by teto** [comment]
I am using the version from https://github.com/NixOS/nixpkgs/pull/172648 which is 0.1.0. Would be interested in testing a 0.1.1 without the bug :) ? I tried running with -v but it didn't change the output

**Comment by elizagamedev** [comment]
Ah, right, I thought your name seemed familiar 🙂 I'll publish a 1.1 sometime tonight with the fixes

**Comment by elizagamedev** [comment]
Okay, v1.1 published and the corresponding nixpkgs PR has been updated as well.

**Comment by teto** [comment]
I've tried again with v1.1 and same behavior:
```
mujmap -C maildir sync -v
error: Could not sync mail: Could not open remote session: Could not open session at https://jmap.fastmail.com:443/.well-known/jmap: https://jmap.fastmail.com/session: status code 401 (redirected from https://jmap.fastmail.com/.well-known/jmap)
```
Is there anything special to do ?
I've replaced my passworkd_command with `password_command = "cat /home/teto/password"` to be sure.
The verbose flag doesn't change the output for me so it's hard to know what's going on.

**Comment by elizagamedev** [comment]
Oops, I think it truly is the newline at the end. I thought *for sure* I had stripped it, because `pass` prints a newline when you run it interactively, but apparently if it detects that its not writing to a TTY it doesn't emit a newline, so I never caught it. 😓

v1.2 incoming

**Comment by elizagamedev** [comment]
Actually, I'm not so sure that's the problem after all. Running v1.1 with my password_command as `cat ~/password` where ~/password has a newline actually does authenticate me with the Fastmail server just fine.

**Comment by elizagamedev** [comment]
I wonder if perhaps the issue here is a typo in your username, or perhaps if you have multiple aliases, Fastmail is particularly picky about picking one of them when authenticating with the "third party" passwords. Do the Fastmail login logs show the failed authorization attempts?

I'm reluctant to add logging which is too detailed when it comes to authentication failures to not leak information, but you could add this patch as a nix override to print the authorization header to stdout:

```patch
diff --git a/src/remote.rs b/src/remote.rs
index aba0186..8460e61 100644
--- a/src/remote.rs
+++ b/src/remote.rs
@@ -98,6 +98,7 @@ impl HttpWrapper {
             "Basic {}",
             base64::encode(format!("{}:{}", safe_username, password))
         );
+        println!("AUTHORIZATION: {authorization}");
         let agent = ureq::AgentBuilder::new()
             .redirect_auth_headers(ureq::RedirectAuthHeaders::SameHost)
             .timeout(Duration::from_secs(timeout))
```

**Comment by teto** [comment]
the logging doesn't even show what file it is reading from.
I've tried mixing credentials (normal password/imap one) and I've no issue with isync or meli, just with mujmap.

I tried to go to https://jmap.fastmail.com/.well-known/jmap where I tried to input my credentials but it seems to do nothing and reloads the prompt for credentials. Do you manage to get something else ?

**Comment by teto** [comment]
arf I've checked the third-party configuration following up your suggestion to look at logs and I was using the credentials for mbsync that was not allowed to access the jmap component it seems, I will try to generate a newset of credentials

**Comment by elizagamedev** [comment]
When I visit in Firefox, I enter my credentials and I get the correct `Session` JSON response. I think there might be an issue with the ones you're using.

You said "normal password/imap one", but Fastmail requires a specific JMAP password which isn't compatible with the IMAP logins. Are you accidentally using an IMAP specific one instead?

**Comment by elizagamedev** [comment]
I used the information from this troubleshooting section to create a new troubleshooting guide in the readme for future users. I'll close this issue once the symlink bug is resolved.

**Comment by teto** [comment]
creating a 3rd party with JMAP access fixed it, sorry for the trouble /o\ I had forgotten that permissions were this detailed on fastmail.

I think this part could use an folder hierarchy example:

> In the directory that you want to use as the maildir for a specific mujmap instance, place a mujmap.toml file ([example](https://github.com/elizagamedev/mujmap/blob/main/mujmap.toml.example)). This directory must be a subdirectory of the notmuch root directory.

I've first hit: 
```
error: Could not sync mail: Could not index local emails: Could not execute notmuch query `path:"/**"': A Xapian exception occurred
```
and then moved the mujmap to the correct location. It is now doing
```
Retrieving metadata... (10665 possibly changed)
Downloading new mail...
```

crossing fingers :)

**Comment by teto** [comment]
a first run failed: 
```
n: status code 403
[2022-05-18T10:27:53Z WARN  mujmap::sync] Download error on try 4, retrying: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/Gc359a2b85a3f72ab3088abeb43ec8dd19d97ef5d/Gc359a2b85a3f72ab3088abeb43ec8dd19d97ef5d?type=text%2Fplain: status code 403
[2022-05-18T10:27:53Z WARN  mujmap::sync] Download error on try 4, retrying: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/Ge24e284b6718f55fad5186cc486c862cd7814eb1/Ge24e284b6718f55fad5186cc486c862cd7814eb1?type=text%2Fplain: status code 403
[2022-05-18T10:27:53Z WARN  mujmap::sync] Download error on try 4, retrying: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/G70ba4ea9c04feded04f64f2831c1f3164126c03b/G70ba4ea9c04feded04f64f2831c1f3164126c03b?type=text%2Fplain: status code 403
[2022-05-18T10:27:53Z WARN  mujmap::sync] Download error on try 4, retrying: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/Gac3cd90130453b687f2e6262599f5a878d0bf19c/Gac3cd90130453b687f2e6262599f5a878d0bf19c?type=text%2Fplain: status code 403
[2022-05-18T10:27:53Z WARN  mujmap::sync] Download error on try 4, retrying: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/G0e9b542a7c818b2fec523f5cd854bd86e8073da6/G0e9b542a7c818b2fec523f5cd854bd86e8073da6?type=text%2Fplain: status code 403
[2022-05-18T10:27:53Z WARN  mujmap::sync] Download error on try 4, retrying: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/Ge360d347a5c907fa27ec14fc203f039f4143fe2e/Ge360d347a5c907fa27ec14fc203f039f4143fe2e?type=text%2Fplaierror: Could not sync mail: Could not download email from remote: Could not read Email blob from server: https://www.fastmailusercontent.com/publicjmap/download/u9ed140f1/G77b05e3f57c9cc1dcc9dcf420589eeacfcb808a0/G77b05e3f57c9cc1dcc9dcf420589eeacfcb808a0?type=text%2Fplain: status code 403
```
maybe because of some throttling ? rerunning mujmap fixed it
```

✖1 ❯ mujmap -C ~/maildir/fastmail -v sync
[2022-05-18T10:37:11Z WARN  mujmap::sync] Could not read mujmap state file `/home/teto/maildir/fastmail/mujmap.state.json': No such file or directory (os error 2)
Retrieving metadata... (10668 possibly changed)
Downloading new mail...
███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 668/668
Applying changes to notmuch database... (10668 new, 10668 changed, 0 destroyed)
error: Could not sync mail: Could not rename mail file from `/home/teto/.cache/mujmap/!home!teto!maildir!fastmail!cur!Mb1bf4816142649a588a0dfc8.Gb1bf4816142649a588a0dfc89420547d88cf4d8d' to `/home/teto/maildir/fastmail/cur/Mb1bf4816142649a588a0dfc8.Gb1bf4816142649a588a0dfc89420547d88cf4d8d': Invalid cross-device link (os error 18)
mujmap -C ~/maildir/fastmail -v sync  39.23s user 15.53s system 12% cpu 7:05.28 total
jinko-doc on  master took 7m5s 
✖1 ❯ mujmap -C ~/maildir/fastmail -v sync
[2022-05-18T10:45:14Z WARN  mujmap::sync] Could not read mujmap state file `/home/teto/maildir/fastmail/mujmap.state.json': No such file or directory (os error 2)
Retrieving metadata... (10671 possibly changed)
Downloading new mail...
███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████ 3/3
mujmap was unable to read the notmuch database revision (stored in
mujmap.state.json) since the last time it was run. As a result, it cannot
determine the changes made in the local database since the last time a
synchronization was performed.

Would you like to potentially discard edits to the notmuch database and replace
them with the JMAP server's changes? This will not affect any notmuch messages
not managed by mujmap or other maildirs managed by mujmap.

Alternatively, you may cancel and attempt to resolve this manually by adding a
`notmuch_revision' to a specific notmuch database revision number in the
`mujmap.state.json' file.

Continue? (y/N)
```
so my error `Invalid cross-device link (os error 18)` is due to mails being on a different device (both are ext4) . I dont mind moving my mail to my main partition to work around it but I thought I would report it anyway in case there is a possible mujmap fix.

> This will not affect any notmuch messages not managed by mujmap or other maildirs managed by mujmap.

I find this sentance hard to parse (maybe it's easier for native English but too many negations)

**Comment by elizagamedev** [comment]
I will make a separate issue regarding the cache being on a separate device. You can set XDG_CACHE_DIR when running mujmap to a location on the same device as your database as a workaround.

Sorry about the confusing error message, but for the initial sync it's fine to enter `y`. I'm not sure why that message was triggered in this case, though.

**Comment by elizagamedev** [comment]
OK, that message is appearing due to an incorrect assumption baked into the code. Making another issue.

**Comment by elizagamedev** [comment]
I'm gonna close this since you were able to get everything working. The symlink issue is tracked in #16.

**Comment by teto** [comment]
sure thanks for the help throughout

**Comment by teto** [comment]
I think I still have the issue with mujmap 2.0 . I have mujmap working with the same (nix) config on a laptop where everything is on the same disk. But on my other computer where the mail is on a separate disk, I get
```
$ mujmap sync -v
(sorry for the terminal control characters)
[m[90m[[39m2022-06-27T22:31:59Z [34mDEBUG[39m mujmap::sync[90m][39m Making symlink from `/home/teto/.cache/mujmap/!mnt!ext!maildir!fastmail!cur!Meecf8641f90b8c68c046b423.Geecf8641f90b8c68c046b4238ddf45446a9ec50e' to `/mnt/ext/maildir/fastmail/cur/Meecf8641f90b8c68c046b423.Geecf8641f90b8c68c046b4238ddf45446a9ec50e'
[m[90m[[39m2022-06-27T22:31:59Z [34mDEBUG[39m mujmap::local[90m][39m Adding new email: NewEmail { remote_email: Email { id: Id("M90bb0cfb7d82a893cbfc2cfc"), blob_id: Id("G90bb0cfb7d82a893cbfc2cfc32a97be9121f1ed7"), keywords: {Seen}, mailbox_ids: {Id("24e4de2e-a4e2-4cd9-822e-08efbf5efab9")}, tags: {} }, cache_path: "/home/teto/.cache/mujmap/!mnt
[m!ext!maildir!fastmail!cur!M90bb0cfb7d82a893cbfc2cfc.G90bb0cfb7d82a893cbfc2cfc32a97be9121f1ed7", maildir_path: "/mnt/ext/maildir/fastmail/cur/M90bb0cfb7d82a893cbfc2cfc.G90bb0cfb7d82a893cbfc2cfc32a97be9121f1ed7" }
[m[90m[[39m2022-06-27T22:31:59Z [34mDEBUG[39m mujmap::sync[90m][39m Removing symlink `/mnt/ext/maildir/fastmail/cur/M90bb0cfb7d82a893cbfc2cfc.G90bb0cfb7d82a893cbfc2cfc32a97be9121f1ed7'
[m[90m[[39m2022-06-27T22:31:59Z [34mDEBUG[39m mujmap::sync[90m][39m Removing symlink `/mnt/ext/maildir/fastmail/cur/Md1189b896c47f667df23c564.Gd1189b896c47f667df23c564c8af7adaad07b156'
```
my generated mujmap.toml (from home-manager) 
```
auto_create_new_mailboxes = true
config_dir = "/home/teto/maildir"
fqdn = "fastmail.com"
password_command = "/nix/store/ry5a1qkccgbcdwnh825zgnhifzdi4mi0-pass-show/bin/pass-show perso/fastmail_mc_jmap"
username = "REDACTED@fastmail.com"

[tags]
deleted = "deleted"
directory_separator = "/"
important = "important"
inbox = "inbox"
lowercase = false
phishing = "phishing"
sent = "sent"
spam = "spam"
```

I've also tried setting `export XDG_CACHE_DIR=$PWD` before running mujmap but I get the same output which I find weird.
```
mujmap --version
mujmap 0.2.0
```

**Comment by elizagamedev** [comment]
Have you tried setting the new `cache_dir` config option to point to a directory on the same device as your mail directory?

**Comment by teto** [comment]
ha nope, I had set config_dir thinking it would act as cache_dir. Anyway I've moved all my mail from a mechanical drive to my main drive (SSD) and the speed bump when tagging mail is priceless so I wont be able to test this anymore. Sorry for the noise.
